File: phases/09-computer-networks/19-p2p-networks-kademlia-bittorrent/code/main.py
"""
P2P Networks — Kademlia, BitTorrent
Phase 09 — Computer Networks

A Kademlia DHT simulation with routing table, XOR distance, iterative lookup,
store/retrieve, and BitTorrent rarest-first piece selection with tit-for-tat
choking/unchoking.
"""

import hashlib
import random
from typing import Optional


class Node:
    """A Kademlia node identified by a 160-bit (20-byte) ID.

    The ID is typically SHA-1 of a public key or random seed.
    """

    def __init__(self, node_id: bytes = None, ip: str = "127.0.0.1", port: int = 0):
        self.id: bytes = node_id or hashlib.sha1(random.randbytes(32)).digest()
        self.ip: str = ip
        self.port: int = port

    def distance(self, other: "Node") -> bytes:
        """XOR distance between this node and another."""
        return bytes(a ^ b for a, b in zip(self.id, other.id))

    def __eq__(self, other):
        return isinstance(other, Node) and self.id == other.id

    def __hash__(self):
        return hash(self.id)

    def __repr__(self):
        return f"Node({self.id[:8].hex()}...@{self.ip}:{self.port})"


def xor_distance_bytes(a: bytes, b: bytes) -> bytes:
    """Compute XOR distance between two byte strings of equal length."""
    if len(a) != len(b):
        raise ValueError(f"ID length mismatch: {len(a)} vs {len(b)}")
    return bytes(x ^ y for x, y in zip(a, b))


def bit_prefix_length(a: bytes, b: bytes) -> int:
    """Count the number of leading bits that are equal in two byte strings.

    This determines which k-bucket a node belongs to relative to the
    local node. Returns an integer in [0, 160] for 20-byte IDs.
    """
    diff = xor_distance_bytes(a, b)
    for i, byte in enumerate(diff):
        if byte != 0:
            return i * 8 + (8 - byte.bit_length())
    return len(a) * 8


class KBucket:
    """A k-bucket stores up to k nodes within a specific XOR distance range.

    Nodes are ordered by last-seen time (most recent at the tail).
    When the bucket is full, the least-recently-seen node (at the head)
    is evicted — but only after a ping confirms it is unresponsive.
    """

    def __init__(self, k: int = 20):
        self.k = k
        self.nodes: list[Node] = []

    def add(self, node: Node):
        """Add a node to the bucket, or refresh it if already present.

        If the bucket is full, the implementation drops the new node.
        A production DHT would ping the head node first.
        """
        if node in self.nodes:
            self.nodes.remove(node)
            self.nodes.append(node)
        elif len(self.nodes) < self.k:
            self.nodes.append(node)

    def remove(self, node: Node):
        """Remove a node from the bucket if present."""
        if node in self.nodes:
            self.nodes.remove(node)

    def get_nodes(self) -> list[Node]:
        return list(self.nodes)

    def __len__(self):
        return len(self.nodes)

    def __repr__(self):
        return f"KBucket(k={self.k}, nodes={len(self.nodes)})"


class RoutingTable:
    """A Kademlia routing table organized as 160 k-buckets.

    Bucket i contains nodes that share the first i bits with the
    local node's ID but differ at bit i. This forms a binary tree
    where each leaf is a bucket covering a XOR distance range.
    """

    def __init__(self, local_id: bytes, k: int = 20, num_buckets: int = 160):
        if len(local_id) != 20:
            raise ValueError("Kademlia node ID must be 20 bytes (160 bits)")
        self.local_id = local_id
        self.k = k
        self.buckets = [KBucket(k) for _ in range(num_buckets)]

    def _bucket_index(self, node_id: bytes) -> int:
        """Return the k-bucket index for a given node ID."""
        return min(bit_prefix_length(self.local_id, node_id), len(self.buckets) - 1)

    def add(self, node: Node):
        """Insert a node into the appropriate k-bucket."""
        if node.id == self.local_id:
            return
        idx = self._bucket_index(node.id)
        self.buckets[idx].add(node)

    def remove(self, node: Node):
        """Remove a node from its k-bucket."""
        if node.id == self.local_id:
            return
        idx = self._bucket_index(node.id)
        self.buckets[idx].remove(node)

    def find_closest(self, target_id: bytes, count: int = 20) -> list[Node]:
        """Return up to count nodes closest to target_id by XOR distance.

        The algorithm searches outward from the bucket that would contain
        the target, collecting candidates until we have enough.
        """
        seen: set[bytes] = set()
        candidates: list[tuple[int, Node]] = []

        idx = self._bucket_index(target_id)

        max_offset = len(self.buckets)
        for offset in range(max_offset):
            for sign in (-1, 1):
                neighbor = idx + offset * sign
                if 0 <= neighbor < len(self.buckets):
                    for n in self.buckets[neighbor].get_nodes():
                        if n.id not in seen:
                            seen.add(n.id)
                            d = int.from_bytes(
                                xor_distance_bytes(n.id, target_id), 'big'
                            )
                            candidates.append((d, n))

        candidates.sort(key=lambda x: x[0])
        return [n for _, n in candidates[:count]]

    def bucket_summary(self) -> dict[int, int]:
        """Return a dict mapping bucket index to node count (for diagnostics)."""
        return {i: len(b) for i, b in enumerate(self.buckets) if len(b) > 0}

    def total_nodes(self) -> int:
        return sum(len(b) for b in self.buckets)


class NodeRegistry:
    """Simulated global node registry for DHT lookups.

    In a real Kademlia network, FIND_NODE requests are sent as UDP
    messages to remote peers. This registry simulates the responses
    so we can test the lookup algorithm without a real network.
    """

    def __init__(self):
        self._peers: dict[bytes, list[Node]] = {}

    def register(self, node_id: bytes, known_peers: list[Node]):
        """Record which peers a given node knows about."""
        self._peers[node_id] = known_peers

    def lookup(self, node_id: bytes) -> list[Node]:
        """Return the peers that node_id would return in a FIND_NODE response."""
        return self._peers.get(node_id, [])

    def size(self) -> int:
        return len(self._peers)


class KademliaNode:
    """A Kademlia DHT node supporting lookup, store, and retrieve operations."""

    def __init__(self, node: Node, registry: NodeRegistry):
        self.node = node
        self.routing_table = RoutingTable(node.id)
        self.data: dict[str, str] = {}
        self.registry = registry

    def add_peer(self, peer: Node):
        """Insert a peer into the local routing table."""
        self.routing_table.add(peer)

    def find_node(self, target_id: bytes, alpha: int = 3) -> list[Node]:
        """Iterative Kademlia node lookup.

        Returns the alpha closest nodes to target_id by:
        1. Seeding candidates from the local routing table.
        2. Querying the closest unqueried candidate.
        3. Repeating until no closer nodes are discovered.
        """
        candidates = self.routing_table.find_closest(target_id, max(alpha * 2, 1))
        queried: set[bytes] = set()

        for iteration in range(16):  # safety limit: O(log N) typical
            unqueried = [
                n for n in candidates
                if n.id not in queried and n.id != self.node.id
            ]
            if not unqueried:
                break

            batch = unqueried[:alpha]
            found_closer = False

            for peer in batch:
                queried.add(peer.id)
                responses = self.registry.lookup(peer.id)
                for rn in responses:
                    if rn.id not in queried and rn.id != self.node.id:
                        self.routing_table.add(rn)
                        found_closer = True

            if not found_closer:
                break

            candidates = self.routing_table.find_closest(
                target_id, max(alpha * 2, 1)
            )

        return [n for n in candidates if n.id != self.node.id][:alpha]

    def store(self, key: str, value: str):
        """Store a key-value pair locally."""
        self.data[key] = value

    def find_value(self, key: str, alpha: int = 3) -> Optional[str]:
        """Iterative FIND_VALUE: look up a key's value in the DHT.

        Hashes the key with SHA-1, performs FIND_NODE, and checks
        whether any of the returned peers hold the value.
        """
        key_hash = hashlib.sha1(key.encode()).digest()
        closest = self.find_node(key_hash, alpha)

        for peer in closest:
            responses = self.registry.lookup(peer.id)
            for rn in responses:
                dist = int.from_bytes(
                    xor_distance_bytes(rn.id, key_hash), 'big'
                )
                if dist < 2 ** 30 and key in self.data:
                    return self.data[key]

        return None


class BitTorrentPeer:
    """A BitTorrent peer managing piece availability and choking state.

    Implements rarest-first piece selection and tit-for-tat choking.
    """

    def __init__(self, peer_id: str, total_pieces: int = 100):
        self.peer_id = peer_id
        self.total_pieces = total_pieces
        self.have_pieces: set[int] = set()
        self.upload_rates: dict[str, float] = {}
        self.optimistic_unchoke_peer: Optional[str] = None
        self._opt_unchoke_counter: int = 0

    def add_piece(self, piece_index: int):
        """Mark a piece as acquired."""
        if 0 <= piece_index < self.total_pieces:
            self.have_pieces.add(piece_index)

    def has_piece(self, piece_index: int) -> bool:
        """Check if we already have a piece."""
        return piece_index in self.have_pieces

    def have_fraction(self) -> float:
        """Fraction of total pieces acquired."""
        return len(self.have_pieces) / self.total_pieces

    def rarest_first_selection(
        self, swarm_rarity: dict[int, int], count: int = 5
    ) -> list[int]:
        """Select count missing pieces, ordered by rarity (rarest first).

        Args:
            swarm_rarity: dict mapping piece_index -> number of peers that have it.
            count: maximum number of pieces to return.

        Returns:
            List of piece indices sorted by increasing rarity.
        """
        missing = [
            p for p in range(self.total_pieces)
            if p not in self.have_pieces
        ]
        missing.sort(key=lambda p: swarm_rarity.get(p, 0))
        return missing[:count]

    def compute_unchoke_set(self, upload_slots: int = 4) -> list[str]:
        """Tit-for-tat unchoking algorithm.

        Unchokes the upload_slots peers with the highest upload rate to us.
        Additionally, one optimistic unchoke slot rotates through remaining
        peers every 3 calls to discover new potential upload partners.

        Args:
            upload_slots: number of regular unchoke slots.

        Returns:
            List of peer IDs that are unchoked.
        """
        if not self.upload_rates:
            return []

        sorted_peers = sorted(
            self.upload_rates.items(),
            key=lambda x: x[1],
            reverse=True,
        )
        unchoked = [pid for pid, _ in sorted_peers[:upload_slots]]

        remaining = [pid for pid in self.upload_rates if pid not in unchoked]
        if remaining:
            self._opt_unchoke_counter += 1
            if self._opt_unchoke_counter % 3 == 0 or self.optimistic_unchoke_peer is None:
                self.optimistic_unchoke_peer = random.choice(remaining)
            if self.optimistic_unchoke_peer not in unchoked:
                unchoked.append(self.optimistic_unchoke_peer)

        return unchoked

    def set_upload_rate(self, peer_id: str, rate_kbps: float):
        """Record a peer's upload rate to us (KB/s)."""
        self.upload_rates[peer_id] = rate_kbps


def main():
    print("=" * 60)
    print("P2P Networks \u2014 Kademlia DHT + BitTorrent")
    print("=" * 60)

    # ==================================================================
    #  Part 1: Kademlia DHT
    # ==================================================================
    print("\n[1] KADEMLIA DHT\n")

    registry = NodeRegistry()

    # Create a bootstrap node
    bootstrap = Node(ip="127.0.0.1", port=8000)
    kademlia = KademliaNode(bootstrap, registry)

    # Create 50 peer nodes and register them with bootstrap's routing table
    peers: list[Node] = []
    for i in range(50):
        peer = Node(ip="127.0.0.1", port=8001 + i)
        kademlia.add_peer(peer)
        peers.append(peer)
    registry.register(bootstrap.id, peers)

    # Give each peer knowledge of a few others (simulating partial network views)
    for i, peer in enumerate(peers):
        known = [
            p for j, p in enumerate(peers)
            if j != i and j % 4 == 0
        ]
        registry.register(peer.id, known)

    # Routing table diagnostics
    summary = kademlia.routing_table.bucket_summary()
    print(f"Bootstrap node: {bootstrap.id.hex()[:16]}...")
    print(f"Routing table: {kademlia.routing_table.total_nodes()} nodes in {len(summary)} buckets")
    if summary:
        print(f"  Buckets with nodes: {sorted(summary.keys())[:5]}...")

    # FIND_NODE for a random target
    target = Node()
    print(f"\nLookup target: {target.id.hex()[:16]}...")
    closest = kademlia.find_node(target.id)
    print(f"Found {len(closest)} closest nodes:")
    for c in closest:
        d = int.from_bytes(xor_distance_bytes(c.id, target.id), 'big')
        print(f"  {c.id.hex()[:16]}...  distance=0x{d:x}")

    # Verify XOR distance properties
    d1 = int.from_bytes(xor_distance_bytes(bootstrap.id, target.id), 'big')
    d2 = int.from_bytes(xor_distance_bytes(target.id, bootstrap.id), 'big')
    assert d1 == d2, f"XOR distance not symmetric: {d1} != {d2}"
    print(f"\nXOR symmetry: dist(A,B) = {d1} == dist(B,A) = {d2} (OK)")

    zero = xor_distance_bytes(bootstrap.id, bootstrap.id)
    assert all(b == 0 for b in zero), "XOR distance from self to self should be zero"
    print(f"XOR identity: dist(A,A) = 0 (OK)")

    # STORE and FIND_VALUE
    print("\n--- STORE / FIND_VALUE ---")
    kademlia.store("course", "computer-networks")
    kademlia.store("topic", "kademlia-dht")
    result = kademlia.find_value("course")
    print(f"FIND_VALUE('course') = {result}")
    result = kademlia.find_value("topic")
    print(f"FIND_VALUE('topic')  = {result}")
    result = kademlia.find_value("nonexistent")
    print(f"FIND_VALUE('nonexistent') = {result}")

    # ==================================================================
    #  Part 2: BitTorrent Piece Selection
    # ==================================================================
    print("\n[2] BITTORRENT PIECE SELECTION\n")

    # Build a swarm rarity map: 100 pieces, random availability
    swarm_rarity = {i: random.randint(1, 50) for i in range(100)}

    bt_peer = BitTorrentPeer("downloader", total_pieces=100)

    # Seed random pieces as already owned
    initial_pieces = set(random.sample(range(100), 35))
    for p in initial_pieces:
        bt_peer.add_piece(p)

    print(f"Peer has {len(bt_peer.have_pieces)}/100 pieces ({bt_peer.have_fraction()*100:.0f}%)")
    selected = bt_peer.rarest_first_selection(swarm_rarity, count=8)
    print(f"\nRarest-first selection (top 8):")
    for p in selected:
        print(f"  Piece {p:3d}  rarity={swarm_rarity.get(p, 0):2d} peers have it")

    # Verify selection correctness
    missing = [p for p in range(100) if p not in bt_peer.have_pieces]
    missing_rarities = sorted([swarm_rarity[p] for p in missing])
    selected_rarities = [swarm_rarity[p] for p in selected]
    assert selected_rarities == missing_rarities[:len(selected)], \
        "Rarest-first order does not match expected rarity order"
    print(f"\nCorrectness: rarest selection verified (OK)")

    # ==================================================================
    #  Part 3: BitTorrent Tit-for-Tat Choking
    # ==================================================================
    print("\n[3] BITTORRENT TIT-FOR-TAT CHOKING\n")

    choker = BitTorrentPeer("choker", total_pieces=100)

    # Simulate 8 leechers with varying upload speeds
    leecher_ids = [f"leecher-{i}" for i in range(8)]
    leecher_rates = [random.uniform(50.0, 1000.0) for _ in range(8)]

    for lid, rate in zip(leecher_ids, leecher_rates):
        choker.set_upload_rate(lid, rate)

    unchoked = choker.compute_unchoke_set(upload_slots=4)

    print(f"Upload slots: 4 regular + 1 optimistic")
    print(f"{'Peer':<14s} {'Rate (KB/s)':<12s} {'Status':<12s}")
    print("-" * 40)
    for pid, rate in sorted(choker.upload_rates.items(),
                            key=lambda x: x[1], reverse=True):
        status = "UNCHOKED" if pid in unchoked else "CHOKED"
        print(f"{pid:<14s} {rate:<12.1f} [{status:<9s}]")

    print(f"\nOptimistic unchoke: {choker.optimistic_unchoke_peer}")

    # Simulate unchoke stability over multiple rounds
    print("\n--- Unchoke stability over 6 rounds ---")
    round_results: list[list[str]] = []
    for rnd in range(6):
        unchoked = choker.compute_unchoke_set(4)
        round_results.append(unchoked)

    # Check optimistic unchoke rotation
    rot_tracker = set()
    for rnd, result in enumerate(round_results):
        opt = choker.optimistic_unchoke_peer
        rot_tracker.add(opt)
        print(f"  Round {rnd + 1}: {len(result)} unchoked, optimistic={opt}")
    print(f"  Optimistic candidates over 6 rounds: {len(rot_tracker)} unique")

    print("\nDone.")


if __name__ == "__main__":
    main()

File: phases/09-computer-networks/19-p2p-networks-kademlia-bittorrent/code/main.rs
//! P2P Networks — Kademlia, BitTorrent
//! Phase 09 — Computer Networks
//!
//! A Kademlia DHT routing table with 160-bit node IDs, XOR distance
//! computation, k-bucket storage, and closest-node lookup.
//! Includes BitTorrent rarest-first piece selection and tit-for-tat
//! choking.
//!
//! Compile with: rustc main.rs && ./main

use std::collections::HashSet;

/// ------------------------------------------------------------------
///  Kademlia Core
/// ------------------------------------------------------------------

/// A Kademlia node identified by a 160-bit (20-byte) ID.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct Node {
    id: [u8; 20],
    ip: [u8; 4],
    port: u16,
}

impl Node {
    fn new(id: [u8; 20], ip: [u8; 4], port: u16) -> Self {
        Node { id, ip, port }
    }

    /// XOR distance between this node and another: id_a ^ id_b.
    fn xor_distance(&self, other: &Node) -> [u8; 20] {
        let mut dist = [0u8; 20];
        for i in 0..20 {
            dist[i] = self.id[i] ^ other.id[i];
        }
        dist
    }
}

/// Count leading zero bits in a 160-bit array.
/// Returns a value in [0, 160] indicating how many leading bits are zero.
fn leading_zero_bits(a: &[u8; 20]) -> usize {
    for (i, &byte) in a.iter().enumerate() {
        if byte != 0 {
            return i * 8 + (byte.leading_zeros() as usize);
        }
    }
    160
}

/// XOR distance as a u128 (using the first 16 bytes).
fn xor_distance_u128(a: &[u8; 20], b: &[u8; 20]) -> u128 {
    let mut d = [0u8; 16];
    for i in 0..16 {
        d[i] = a[i] ^ b[i];
    }
    u128::from_be_bytes(d)
}

/// A k-bucket holds up to K nodes sorted by last-seen recency.
#[derive(Debug, Clone)]
struct KBucket<const K: usize = 20> {
    nodes: Vec<Node>,
}

impl<const K: usize> KBucket<K> {
    fn new() -> Self {
        KBucket {
            nodes: Vec::with_capacity(K),
        }
    }

    /// Add a node. Moves existing node to tail (most recent).
    /// If full, the new node is dropped (production DHTs would
    /// ping the head node first to check liveness).
    fn add(&mut self, node: Node) {
        if let Some(pos) = self.nodes.iter().position(|n| n.id == node.id) {
            self.nodes.remove(pos);
            self.nodes.push(node);
        } else if self.nodes.len() < K {
            self.nodes.push(node);
        }
    }

    fn remove(&mut self, node: &Node) {
        self.nodes.retain(|n| n.id != node.id);
    }

    fn iter(&self) -> impl Iterator<Item = &Node> {
        self.nodes.iter()
    }

    fn len(&self) -> usize {
        self.nodes.len()
    }
}

/// A Kademlia routing table using 160 k-buckets.
///
/// Bucket `i` covers nodes whose IDs share the first `i` bits with
/// the local ID but differ at bit `i`. This forms a binary tree
/// structure over the 160-bit address space.
struct RoutingTable<const K: usize = 20> {
    local_id: [u8; 20],
    buckets: [KBucket<K>; 160],
}

impl<const K: usize> RoutingTable<K> {
    fn new(local_id: [u8; 20]) -> Self {
        let buckets = [(); 160].map(|_| KBucket::new());
        RoutingTable { local_id, buckets }
    }

    /// Determine which k-bucket a node ID belongs to.
    fn bucket_index(&self, node_id: &[u8; 20]) -> usize {
        let mut diff = [0u8; 20];
        for i in 0..20 {
            diff[i] = self.local_id[i] ^ node_id[i];
        }
        leading_zero_bits(&diff).min(159)
    }

    /// Insert a node into the appropriate bucket.
    fn add(&mut self, node: Node) {
        if node.id == self.local_id {
            return;
        }
        let idx = self.bucket_index(&node.id);
        self.buckets[idx].add(node);
    }

    /// Remove a node from its bucket.
    fn remove(&mut self, node: &Node) {
        let idx = self.bucket_index(&node.id);
        self.buckets[idx].remove(node);
    }

    /// Return the `count` closest known nodes to `target_id`.
    ///
    /// Searches outward from the bucket that would contain the target,
    /// collecting nodes and sorting by XOR distance.
    fn find_closest(&self, target_id: &[u8; 20], count: usize) -> Vec<Node> {
        let idx = self.bucket_index(target_id).min(159);
        let mut candidates: Vec<(u128, Node)> = Vec::new();
        let mut seen: HashSet<[u8; 20]> = HashSet::new();

        let max_offset = 160usize;
        for offset in 0..max_offset {
            for sign in [-1isize, 1isize] {
                let neighbor = idx as isize + offset as isize * sign;
                if neighbor < 0 || neighbor >= 160 {
                    continue;
                }
                let bucket = &self.buckets[neighbor as usize];
                for node in bucket.iter() {
                    if !seen.insert(node.id) {
                        continue;
                    }
                    let dist = xor_distance_u128(&node.id, target_id);
                    candidates.push((dist, node.clone()));
                }
            }
        }

        candidates.sort_by_key(|(dist, _)| *dist);
        candidates.truncate(count);
        candidates.into_iter().map(|(_, n)| n).collect()
    }

    /// Return per-bucket node counts for diagnostics.
    fn bucket_counts(&self) -> Vec<usize> {
        self.buckets.iter().map(|b| b.len()).collect()
    }

    /// Total nodes across all buckets.
    fn total_nodes(&self) -> usize {
        self.buckets.iter().map(|b| b.len()).sum()
    }
}

/// ------------------------------------------------------------------
///  BitTorrent Peer
/// ------------------------------------------------------------------

/// A BitTorrent peer with rarest-first piece selection
/// and tit-for-tat choking.
struct BitTorrentPeer {
    peer_id: String,
    total_pieces: usize,
    have_pieces: Vec<bool>,
    upload_rates: Vec<(String, f64)>,
}

impl BitTorrentPeer {
    fn new(peer_id: &str, total_pieces: usize) -> Self {
        BitTorrentPeer {
            peer_id: peer_id.to_string(),
            total_pieces,
            have_pieces: vec![false; total_pieces],
            upload_rates: Vec::new(),
        }
    }

    /// Mark a piece as acquired.
    fn have(&mut self, piece: usize) {
        if piece < self.total_pieces {
            self.have_pieces[piece] = true;
        }
    }

    /// Return missing pieces sorted by rarity (rarest first).
    fn rarest_first(&self, swarm_rarity: &[usize]) -> Vec<usize> {
        let mut missing: Vec<(usize, usize)> = self
            .have_pieces
            .iter()
            .enumerate()
            .filter(|(_, &have)| !have)
            .map(|(i, _)| {
                let rarity = if i < swarm_rarity.len() {
                    swarm_rarity[i]
                } else {
                    0
                };
                (rarity, i)
            })
            .collect();
        missing.sort_by_key(|&(rarity, _)| rarity);
        missing.into_iter().map(|(_, i)| i).collect()
    }

    /// Return fraction of pieces owned.
    fn have_fraction(&self) -> f64 {
        let owned = self.have_pieces.iter().filter(|&&h| h).count();
        owned as f64 / self.total_pieces as f64
    }

    /// Set a peer's upload rate to us (in KB/s).
    fn set_upload_rate(&mut self, peer_id: &str, rate: f64) {
        if let Some(pos) = self.upload_rates.iter().position(|(id, _)| id == peer_id) {
            self.upload_rates[pos].1 = rate;
        } else {
            self.upload_rates.push((peer_id.to_string(), rate));
        }
    }

    /// Compute unchoke set: peers with highest upload rate fill the slots.
    fn compute_unchoke_set(&self, slots: usize) -> Vec<String> {
        let mut sorted = self.upload_rates.clone();
        sorted.sort_by(|a, b| {
            b.1.partial_cmp(&a.1)
                .unwrap_or(std::cmp::Ordering::Equal)
        });
        sorted
            .into_iter()
            .take(slots)
            .map(|(id, _)| id)
            .collect()
    }
}

/// ------------------------------------------------------------------
///  Main
/// ------------------------------------------------------------------

fn main() {
    println!("========================================");
    println!("P2P Networks \u2014 Kademlia + BitTorrent");
    println!("========================================");

    // ==============================================================
    // Part 1: Kademlia Routing Table
    // ==============================================================
    println!("\n--- Kademlia Routing Table ---\n");

    let local_id: [u8; 20] = [
        0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99,
        0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff, 0x00, 0x11, 0x22, 0x33,
    ];
    let local = Node::new(local_id, [10, 0, 0, 1], 8000);
    let mut table = RoutingTable::<20>::new(local.id);

    // Create 60 nodes with predictable IDs and add them to the table
    let ip: [u8; 4] = [10, 0, 0, 0];
    for i in 0..60 {
        let mut nid = [0u8; 20];
        nid[0] = (i.wrapping_mul(37) & 0xff) as u8;
        nid[1] = (i.wrapping_mul(73) & 0xff) as u8;
        nid[2] = (i.wrapping_mul(151) & 0xff) as u8;
        nid[19] = i as u8;
        let node = Node::new(nid, ip, 9000 + i as u16);
        table.add(node);
    }
    println!("Added 60 nodes to routing table.");
    let counts = table.bucket_counts();
    let occupied: usize = counts.iter().filter(|&&c| c > 0).count();
    println!(
        "Total nodes in routing table: {} ({} occupied buckets)",
        table.total_nodes(),
        occupied
    );

    // Display buckets with most nodes
    let mut bucket_stats: Vec<(usize, usize)> =
        counts.iter().enumerate().filter(|(_, &c)| c > 0).map(|(i, &c)| (i, c)).collect();
    bucket_stats.sort_by_key(|&(_, c)| std::cmp::Reverse(c));
    println!("Top 5 most populated buckets:");
    for (idx, count) in bucket_stats.iter().take(5) {
        println!("  Bucket {:>3}: {} nodes", idx, count);
    }

    // FIND_CLOSEST lookup
    let target_id: [u8; 20] = [
        0xff, 0xee, 0xdd, 0xcc, 0xbb, 0xaa, 0x99, 0x88, 0x77, 0x66,
        0x55, 0x44, 0x33, 0x22, 0x11, 0x00, 0xff, 0xee, 0xdd, 0xcc,
    ];
    let target = Node::new(target_id, [10, 0, 0, 2], 8001);

    let closest = table.find_closest(&target.id, 5);
    println!("\nClosest 5 nodes to target");
    println!("  Target ID: {:02x}{:02x}{:02x}{:02x}...", target_id[0], target_id[1], target_id[2], target_id[3]);
    for (i, n) in closest.iter().enumerate() {
        let dist = local.xor_distance(n);
        println!(
            "  {}. {:02x}{:02x}{:02x}{:02x}...  distance=0x{:02x}{:02x}{:02x}{:02x}...",
            i + 1,
            n.id[0], n.id[1], n.id[2], n.id[3],
            dist[0], dist[1], dist[2], dist[3]
        );
    }

    // Verify XOR distance symmetry
    let d1 = local.xor_distance(&target);
    let d2 = target.xor_distance(&local);
    assert_eq!(d1, d2, "XOR distance must be symmetric");
    println!("\nXOR symmetry: dist(A,B) == dist(B,A) (OK)");

    // Verify XOR self-distance is zero
    let self_dist = local.xor_distance(&local);
    assert!(self_dist.iter().all(|&b| b == 0), "Self-distance must be 0");
    println!("XOR identity: dist(A,A) == 0 (OK)");

    // ==============================================================
    // Part 2: BitTorrent Piece Selection
    // ==============================================================
    println!("\n--- BitTorrent Piece Selection ---\n");

    // Build a swarm rarity map: 100 pieces with pseudo-random availability
    // rarity = number of peers that have the piece (higher = more common)
    let mut swarm_rarity = vec![0usize; 100];
    for i in 0..100 {
        swarm_rarity[i] = ((i.wrapping_mul(7) + 13) % 50) + 1;
    }

    let mut bt_peer = BitTorrentPeer::new("peer-a", 100);

    // Seed with 30 owned pieces
    let seeded: [usize; 30] = [
        3, 7, 12, 15, 18, 22, 27, 31, 34, 38, 42, 45, 49, 51, 55,
        58, 62, 66, 70, 73, 77, 80, 84, 87, 89, 91, 94, 96, 98, 99,
    ];
    for &p in &seeded {
        bt_peer.have(p);
    }

    println!(
        "Peer has {}/100 pieces ({:.0}%)",
        bt_peer.have_pieces.iter().filter(|&&h| h).count(),
        bt_peer.have_fraction() * 100.0
    );

    let rarest_order = bt_peer.rarest_first(&swarm_rarity);
    println!("Rarest-first selection (top 10):");
    for &p in rarest_order.iter().take(10) {
        println!("  Piece {:>3}  rarity={}", p, swarm_rarity[p]);
    }

    // Verify rarest-first correctness
    let missing_rarities: Vec<usize> = bt_peer
        .have_pieces
        .iter()
        .enumerate()
        .filter(|(_, &have)| !have)
        .map(|(i, _)| swarm_rarity[i])
        .collect();
    let mut sorted_missing = missing_rarities.clone();
    sorted_missing.sort();
    let selected_rarities: Vec<usize> =
        rarest_order.iter().take(10).map(|&p| swarm_rarity[p]).collect();
    assert_eq!(
        selected_rarities,
        sorted_missing[..10.min(sorted_missing.len())],
        "Rarest-first order incorrect"
    );
    println!("  Rarest-first correctness: verified (OK)");

    // ==============================================================
    // Part 3: Tit-for-Tat Choking
    // ==============================================================
    println!("\n--- Tit-for-Tat Choking ---\n");

    let mut choker = BitTorrentPeer::new("choker", 100);

    // 3 fast peers, 3 medium, 2 slow
    choker.set_upload_rate("fast-a", 950.0);
    choker.set_upload_rate("fast-b", 880.0);
    choker.set_upload_rate("fast-c", 720.0);
    choker.set_upload_rate("medium-a", 510.0);
    choker.set_upload_rate("medium-b", 430.0);
    choker.set_upload_rate("medium-c", 390.0);
    choker.set_upload_rate("slow-a", 120.0);
    choker.set_upload_rate("slow-b", 85.0);

    let unchoked = choker.compute_unchoke_set(4);
    println!("Upload slots = 4 (tit-for-tat)");
    println!("{:<12s} {:>10s}  {:<10s}", "Peer", "Rate KB/s", "Status");
    println!("{}", "-".repeat(36));
    for (id, rate) in &choker.upload_rates {
        let status = if unchoked.contains(id) {
            "UNCHOKED"
        } else {
            "CHOKED"
        };
        println!("{:<12s} {:>10.1}  [{:<8s}]", id, rate, status);
    }

    let fast_unchoked: Vec<&String> = unchoked.iter().filter(|id| id.starts_with("fast")).collect();
    assert_eq!(fast_unchoked.len(), 3, "All fast peers should be unchoked");
    println!("\n  All 3 fast peers unchoked: verified (OK)");

    println!("\nDone.");
}
